pymc4/coroutine_model.py

```diff
@@ -0,0 +1,236 @@
+import biwrap
+import functools
+import types
+import pymc4
+from pymc4.scopes import name_scope
+
+
+@biwrap.biwrap
+def model(genfn, *, name=None, keep_auxiliary=True, keep_return=True, method=False):
+    if method:
+        template = ModelTemplate(
+            genfn, name=name, keep_auxiliary=keep_auxiliary, keep_return=keep_return
+        )
+
+        @functools.wraps(genfn)
+        def wrapped(*args, **kwargs):
+            return template(*args, **kwargs)
+
+        return wrapped
+    else:
+        template = ModelTemplate(
+            genfn, name=name, keep_auxiliary=keep_auxiliary, keep_return=keep_return
+        )
+        return template
+
+
+def get_name(default, base_fn, name):
+    if name is None:
+        if default is not None:
+            name = default
+        elif hasattr(base_fn, "name"):
+            name = getattr(base_fn, "name")
+        elif hasattr(base_fn, "__name__"):
+            name = base_fn.__name__
+    return name
+
+
+class ModelTemplate(object):
+    """
+    The `ModelTemplate` object, this keeps a function, responsible for a specific generative process.
+    Generative process is one that samples from prior distributions and interacts
+    them in a user defined way arbitrarily complex.
+
```

Comment:
Suggested re-wording: `ModelTemplate` is a callable object that represents a generative process. A generative process samples from prior distributions and allows them to interact in arbitrarily-complex, user-defined ways.

Reply:
Moreover we say any noarg generator function is a model template. I do not like default args and thinking over a more beautiful solution here